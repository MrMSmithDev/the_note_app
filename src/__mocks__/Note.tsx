import React from 'react';
import type { NoteProps } from '@components/Note/Note';

export default function Note({ date }: NoteProps) {
  return <div data-testid="mock-note">Mock Note for {date}</div>;
}
