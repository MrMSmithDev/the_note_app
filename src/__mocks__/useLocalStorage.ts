export const mockSaveNote = jest.fn();
export const mockRemoveNote = jest.fn();
export const mockRetrieveNotes = jest.fn().mockReturnValue([]);
export const mockLoadYear = jest.fn();
export const mockUpdateNote = jest.fn();

export const useLocalStorage = jest.fn(() => ({
  loadYear: mockLoadYear,
  saveNoteToLocal: mockSaveNote,
  removeNoteFromLocal: mockRemoveNote,
  retrieveNotesFromLocal: mockRetrieveNotes,
  updateNoteToLocal: mockUpdateNote,
}));
