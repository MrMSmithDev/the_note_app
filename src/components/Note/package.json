{
    "name": "note",
    "version": "1.0.0",
    "main": "Note.tsx",
    "dependencies": {
      "react": "^19.0.0"
    }
}