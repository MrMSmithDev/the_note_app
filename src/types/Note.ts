export default interface Note {
  id: number;
  data: string;
}
