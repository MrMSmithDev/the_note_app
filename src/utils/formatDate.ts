import { monthsData } from './dateData';

export default function formatDate(dateString: string): string {
  const [yearStr, monthStr, dateStr] = dateString.split('-');

  // Set date suffix
  const dateNum = parseInt(dateStr, 10);
  let dateSuffix = 'th';

  if (![11, 12, 13].includes(dateNum)) {
    if (dateNum % 10 === 1) dateSuffix = 'st';
    if (dateNum % 10 === 2) dateSuffix = 'nd';
    if (dateNum % 10 === 3) dateSuffix = 'rd';
  }

  // Retrieve month name
  const monthNum = parseInt(monthStr, 10);
  const month = monthsData[monthNum - 1].abbr || 'UNKNOWN';

  return `${dateStr}${dateSuffix} ${month} ${yearStr}`;
}
