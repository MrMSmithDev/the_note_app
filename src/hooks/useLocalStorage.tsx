import React, { createContext, useContext, useState } from 'react';
import type LocalDataStruct from 'src/types/LocalDataStruct';
import Note from 'src/types/Note';

interface LocalStorageContextType {
  currentData: LocalDataStruct;
  saveNoteToLocal: (dateString: string, note: Note) => void;
  removeNoteFromLocal: (dateString: string, noteId: string) => void;
  loadYear: (year: number) => void;
  retrieveNotesFromLocal: (dateString: string) => Note[];
}

interface LocalStorageProviderProps {
  children: React.ReactNode;
}

const LocalStorageContext = createContext<LocalStorageContextType | undefined>(
  undefined
);

const LocalStorageProvider: React.FC<LocalStorageProviderProps> = ({
  children,
}) => {
  const [currentData, setCurrentData] = useState<LocalDataStruct>({});

  function loadYear(year: number) {
    const storedData = localStorage.getItem(year.toString());

    if (storedData) setCurrentData(JSON.parse(storedData));
    else setCurrentData({});
  }

  function saveNoteToLocal(dateString: string, note: Note) {
    const [yearStr, monthStr, dateStr] = dateString.split('-');
    const month = parseInt(monthStr, 10);
    const date = parseInt(dateStr, 10);

    setCurrentData((prev) => {
      const updatedMonthData = {
        ...prev[month],
        [date]: [...(prev[month]?.[date] || []), note],
      };

      const updatedYearData = {
        ...prev,
        [month]: updatedMonthData,
      };

      localStorage.setItem(yearStr, JSON.stringify(updatedYearData));

      return updatedYearData;
    });
  }

  function removeNoteFromLocal(dateString: string, noteId: string) {
    const [yearStr, monthStr, dateStr] = dateString.split('-');
    const month = parseInt(monthStr, 10);
    const date = parseInt(dateStr, 10);

    setCurrentData((prev) => {
      const updatedMonthData = {
        ...prev[month],
        [date]: prev[month]?.[date].filter((note) => note.id !== noteId),
      };

      const updatedYearData = {
        ...prev,
        [month]: updatedMonthData,
      };

      localStorage.setItem(yearStr, JSON.stringify(updatedYearData));

      return updatedYearData;
    });
  }

  function retrieveNotesFromLocal(dateString: string): Note[] {
    const [_, monthStr, dateStr] = dateString.split('-');
    const month = parseInt(monthStr, 10);
    const date = parseInt(dateStr, 10);

    return currentData[month]?.[date] || [];
  }

  return (
    <LocalStorageContext.Provider
      value={{
        currentData,
        saveNoteToLocal,
        removeNoteFromLocal,
        loadYear,
        retrieveNotesFromLocal,
      }}
    >
      {children}
    </LocalStorageContext.Provider>
  );
};

export function useLocalStorage(): LocalStorageContextType {
  const context = useContext(LocalStorageContext);
  if (!context) {
    throw new Error(
      'useLocalStorage must be used within a LocalStorageProvider'
    );
  }
  return context;
}

export default LocalStorageProvider;
