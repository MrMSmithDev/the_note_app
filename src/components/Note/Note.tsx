import CrossIcon from '@components/icons/CrossIcon';
import { useLocalStorage } from '@hooks/useLocalStorage';
import React, { useEffect, useState } from 'react';
import NoteType from 'src/types/Note';
import { v4 as uuidV4 } from 'uuid';
import formatDate from '@utils/formatDate';
import { lightningCssTransform } from 'next/dist/build/swc/generated-native';

interface NoteProps {
  date: string;
}

const Note: React.FC<NoteProps> = ({ date }) => {
  const [currentNewNote, setCurrentNewNote] = useState<string>('');
  const [noteData, setNoteData] = useState<NoteType[]>([]);
  const { saveNoteToLocal, removeNoteFromLocal, retrieveNotesFromLocal } =
    useLocalStorage();

  const token = null;

  useEffect(() => {
    async function loadData() {
      console.log('loading note data');

      // Check if using auth
      if (token) {
        console.log('token');
      } else {
        const result = retrieveNotesFromLocal(date);
        console.log('Retrieve notes');
        setNoteData(result);
      }
    }

    void loadData();
  }, [date]);

  function handleNewNoteChange(e: React.ChangeEvent<HTMLInputElement>) {
    setCurrentNewNote(e.currentTarget.value);
  }

  function createNote(e: React.MouseEvent<HTMLButtonElement>) {
    e.preventDefault();

    const newNote: NoteType = { id: uuidV4(), data: currentNewNote };
    if (token) {
      setTimeout(() => {
        console.log('token');
      }, 5);
    } else {
      saveNoteToLocal(date, newNote);
      setNoteData((prev) => [...prev, newNote]);
    }

    setCurrentNewNote('');
  }

  async function removeNoteData(e: React.MouseEvent<HTMLButtonElement>) {
    e.preventDefault();

    const dataId = e.currentTarget.value;
    removeNoteFromLocal(date, dataId);
    setNoteData((prev) => prev.filter((data) => data.id !== dataId));
  }

  return (
    <div className="bg-background dark:bg-background-dark rounded shadow-md p-5 min-w-[300px] w-min animate-slide-down">
      <h4 className="bg-clip-text text-transparent bg-gradient-to-r from-blue-400 to-blue-600 font-bold font-kanit">
        {formatDate(date)}
      </h4>
      <ul className="my-3 flex flex-col gap-2">
        {noteData.map((note: NoteType) => (
          <li key={note.id} className="flex items-center">
            <p className="text-left dark:text-gray-200">{note.data}</p>
            <button
              onClick={removeNoteData}
              className="ml-auto pl-2 cursor-pointer"
              value={note.id}
            >
              <span className="sr-only">Delete Note</span>
              <CrossIcon className="h-3 w-3" fill="#c30010" />
            </button>
          </li>
        ))}
        {noteData.length <= 0 ? (
          <li className="text-center italic text-gray-400">No Notes</li>
        ) : null}
      </ul>
      <form className="flex gap-2 flex-col md:flex-row justify-center mt-2">
        <label className="sr-only" htmlFor="newNote">
          New Note
        </label>
        <input
          className="p-1 border-2 border-gray-200 dark:border-gray-800 focus:outline-gray-400 dark:focus:outline-gray-600 text-sm dark:text-gray-200 tracking-wide min-w-[300px]"
          type="text"
          id="newNote"
          name="newNote"
          placeholder="New Note"
          value={currentNewNote}
          onChange={handleNewNoteChange}
        />
        <button
          onClick={createNote}
          className="mx-auto w-max bg-blue-500 py-1 px-4 rounded-md font-quicksand font-bold text-gray-100 cursor-pointer hover:animate-jello"
        >
          <span>Add Note</span>
        </button>
      </form>
    </div>
  );
};

export default Note;
