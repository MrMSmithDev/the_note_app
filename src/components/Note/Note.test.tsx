import { render, screen, fireEvent } from '@testing-library/react';
import Note from './Note';
import {
  mockSaveNote,
  mockRemoveNote,
  mockRetrieveNotes,
} from '@__mocks__/useLocalStorage';
import formatDate from '@utils/formatDate';

jest.mock('@hooks/useLocalStorage', () =>
  require('@__mocks__/useLocalStorage')
);

describe('Note component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with date', () => {
    render(<Note date="2025-02-1" />);
    expect(screen.getByText(formatDate('2025-02-1'))).toBeInTheDocument();
  });

  it('calls retrieveNotesFromLocal on mount when no token present', () => {
    render(<Note date="2025-02-1" />);
    expect(mockRetrieveNotes).toHaveBeenCalledWith('2025-02-1');
  });

  it('updates input value when typed into', () => {
    render(<Note date="2025-02-1" />);
    const input = screen.getByPlaceholderText('New Note');
    fireEvent.change(input, { target: { value: 'Test Note' } });
    expect(input).toHaveValue('Test Note');
  });

  it('adds a new note on submit button click and displays it', () => {
    render(<Note date="2025-02-1" />);

    const input = screen.getByPlaceholderText('New Note');
    const addBtn = screen.getByText('Add Note');

    fireEvent.change(input, { target: { value: 'New Test Note' } });
    fireEvent.click(addBtn);

    expect(mockSaveNote).toHaveBeenCalledWith('2025-02-1', expect.any(Object));
    expect(screen.getByText('New Test Note')).toBeInTheDocument();
  });

  it('removes a note on delete button click', () => {
    mockRetrieveNotes.mockReturnValue([{ id: '1', data: 'Test Note' }]);

    render(<Note date="2025-02-1" />);

    const deleteBtn = screen.getByRole('button', { name: 'Delete Note' });
    fireEvent.click(deleteBtn);

    expect(mockRemoveNote).toHaveBeenCalledWith('2025-02-1', '1');
    expect(screen.queryByText('Test Note')).not.toBeInTheDocument();
  });
});
