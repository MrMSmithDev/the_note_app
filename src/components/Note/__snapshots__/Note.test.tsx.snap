// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note component renders correctly to match snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-background dark:bg-background-dark rounded shadow-md p-5 max-h-[80lvh] min-w-[300px] w-min animate-slide-down"
  >
    <h4
      class="bg-clip-text text-transparent bg-gradient-to-r from-blue-400 to-blue-600 font-bold font-kanit"
    >
      5th JAN 2025
    </h4>
    <ul
      class="my-3 flex flex-col gap-2"
    >
      <li
        class="text-center italic text-gray-400"
      >
        No Notes
      </li>
    </ul>
    <form
      class="flex gap-2 flex-col md:flex-row justify-center mt-2"
    >
      <label
        class="sr-only"
        for="newNote"
      >
        New Note
      </label>
      <input
        class="p-1 border-2 border-gray-200 dark:border-gray-800 focus:outline-gray-400 dark:focus:outline-gray-600 text-sm dark:text-gray-200 tracking-wide min-w-[300px]"
        id="newNote"
        name="newNote"
        placeholder="New Note"
        type="text"
        value=""
      />
      <button
        class="py-2 px-4 rounded-md font-quicksand font-bold text-gray-100 mr-auto bg-linear-to-tr from-blue-400 to-blue-600 w-full md:w-max cursor-pointer disabled:opacity-50"
      >
        <span>
          Add Note
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;
